File: app/src/main/java/com/myflightbook/android/ActVisitedAirports.java
/*
	MyFlightbook for Android - provides native access to MyFlightbook
	pilot's logbook
    Copyright (C) 2017-2018 MyFlightbook, LLC

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package com.myflightbook.android;

import android.app.Activity;
import android.app.ProgressDialog;
import android.content.Context;
import android.content.Intent;
import android.os.AsyncTask;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v4.widget.SwipeRefreshLayout;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.EditText;
import android.widget.SimpleExpandableListAdapter;
import android.widget.TextView;

import com.myflightbook.android.WebServices.AuthToken;
import com.myflightbook.android.WebServices.MFBSoap;
import com.myflightbook.android.WebServices.VisitedAirportSvc;

import java.text.DateFormat;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.Locale;
import java.util.Objects;

import Model.MFBUtil;
import Model.VisitedAirport;

public class ActVisitedAirports extends ExpandableListFragment implements MFBMain.Invalidatable {

    private static VisitedAirport[] visitedAirports = null;

    private static class RefreshVisitedAirports extends AsyncTask<Void, Void, MFBSoap> {
        private ProgressDialog m_pd = null;
        Object m_Result = null;
        private AsyncWeakContext<ActVisitedAirports> m_ctxt;

        RefreshVisitedAirports(Context c, ActVisitedAirports ava) {
            super();
            m_ctxt = new AsyncWeakContext<>(c, ava);
        }

        @Override
        protected MFBSoap doInBackground(Void... params) {
            VisitedAirportSvc vas = new VisitedAirportSvc();
            m_Result = vas.VisitedAirportsForUser(AuthToken.m_szAuthToken, m_ctxt.getContext());
            return vas;
        }

        protected void onPreExecute() {
            m_pd = MFBUtil.ShowProgress(m_ctxt.getContext(), m_ctxt.getContext().getString(R.string.prgVisitedAirports));
        }

        protected void onPostExecute(MFBSoap svc) {
            try {
                if (m_pd != null)
                    m_pd.dismiss();
            } catch (Exception ignored) {
            }

            ActVisitedAirports ava = m_ctxt.getCallingActivity();
            if (ava == null || !ava.isAdded() || ava.isDetached() || ava.getActivity() == null)
                return;

            ActVisitedAirports.visitedAirports = (VisitedAirport[]) m_Result;

            if (ActVisitedAirports.visitedAirports == null || svc.getLastError().length() > 0)
                MFBUtil.Alert(ava, ava.getString(R.string.txtError), svc.getLastError());
            else
                ava.populateList();
        }
    }

    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        super.setHasOptionsMenu(true);

        return inflater.inflate(R.layout.expandablelist, container, false);
    }

    @Override
    public void onActivityCreated(Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        MFBMain.registerNotifyDataChange(this);
        MFBMain.registerNotifyResetAll(this);

        Activity a = getActivity();
        assert a != null;
        TextView tvSearch = a.findViewById(R.id.txtSearchProp);
        tvSearch.setHint(R.string.hintSearchVisitedAirports);
        tvSearch.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence charSequence, int i, int i1, int i2) { }

            @Override
            public void onTextChanged(CharSequence charSequence, int i, int i1, int i2) {
                populateList();
            }

            @Override
            public void afterTextChanged(Editable editable) { }
        });

        SwipeRefreshLayout srl = a.findViewById(R.id.swiperefresh);
        srl.setOnRefreshListener(() -> {
            srl.setRefreshing(false);
            refreshAirports();
        });
    }

    public void onResume() {
        super.onResume();
        if (visitedAirports == null)
            new RefreshVisitedAirports(getActivity(), this).execute();
        else
            populateList();
    }

    @Override
    // reuse currency menu here.
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
        inflater.inflate(R.menu.currencymenu, menu);
    }

    void refreshAirports() {
        new RefreshVisitedAirports(getActivity(), this).execute();
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        // Handle item selection
        switch (item.getItemId()) {
            case R.id.menuRefresh:
                refreshAirports();
                return true;
            default:
                return super.onOptionsItemSelected(item);
        }
    }

    private void populateList() {
        if (visitedAirports == null || visitedAirports.length == 0)
            return;

        // This maps the headers to the individual sub-lists.
        HashMap<String, HashMap<String, String>> headers = new HashMap<>();
        HashMap<String, ArrayList<HashMap<String, String>>> childrenMaps = new HashMap<>();

        // Keep a list of the keys in order
        ArrayList<String> alKeys = new ArrayList<>();

        Arrays.sort(visitedAirports);

        // First do "All airports"
        String szKeyLast = getString(R.string.AllAirports);
        alKeys.add(szKeyLast);
        HashMap<String, String> hmAllAirports = new HashMap<>();
        hmAllAirports.put("sectionName", szKeyLast);
        headers.put(szKeyLast, hmAllAirports);

        hmAllAirports = new HashMap<>();
        hmAllAirports.put("Code", "");
        hmAllAirports.put("Name", szKeyLast);
        int cVisits = visitedAirports.length;
        hmAllAirports.put("Visits", getResources().getQuantityString(R.plurals.uniqueAirportsCount, cVisits, cVisits));
        hmAllAirports.put("Distance", "");
        hmAllAirports.put("Position", "-1");
        ArrayList<HashMap<String, String>> alAllAirports = new ArrayList<>();
        alAllAirports.add(hmAllAirports);
        childrenMaps.put(szKeyLast, alAllAirports);

        String szRestrict = ((EditText) Objects.requireNonNull(getActivity()).findViewById(R.id.txtSearchProp)).getText().toString().toUpperCase(Locale.getDefault());

        // slice and dice into headers/first names
        for (int i = 0; i < visitedAirports.length; i++) {
            VisitedAirport va = visitedAirports[i];

            if (szRestrict.length() > 0 && !va.airport.FacilityName.toUpperCase(Locale.getDefault()).contains(szRestrict) && !va.airport.AirportID.toUpperCase(Locale.getDefault()).contains(szRestrict))
                continue;

            // get the first letter for this property as the grouping key
            String szKey = va.airport.FacilityName.substring(0, 1).toUpperCase(Locale.getDefault());
            if (szKey.compareTo(szKeyLast) != 0) {
                alKeys.add(szKey);
                szKeyLast = szKey;
            }

            HashMap<String, String> hmGroups = (headers.containsKey(szKey) ? headers.get(szKey) : new HashMap<>());
            hmGroups.put("sectionName", szKey);
            headers.put(szKey, hmGroups);

            // Get the array-list for that key, creating it if necessary
            ArrayList<HashMap<String, String>> alAirports;
            alAirports = (childrenMaps.containsKey(szKey) ? childrenMaps.get(szKey) : new ArrayList<>());

            HashMap<String, String> hmProperty = new HashMap<>();
            String szDistance = va.airport.Distance > 0 ? String.format(Locale.getDefault(), "(%.1fnm) ", va.airport.Distance) : "";
            String szEarliestVisit = DateFormat.getDateInstance(DateFormat.SHORT).format(va.EarliestDate);
            String szLatestVisit = DateFormat.getDateInstance(DateFormat.SHORT).format(va.LatestDate);
            String szVisits = (va.NumberOfVisits > 1) ?
                    String.format(getString(R.string.vaMultiVisit), va.NumberOfVisits, szEarliestVisit, szLatestVisit) :
                    String.format(getString(R.string.vaSingleVisit), szEarliestVisit);
            hmProperty.put("Code", String.format("%s - ", va.Code));
            hmProperty.put("Name", va.airport.FacilityName);
            hmProperty.put("Distance", szDistance);
            hmProperty.put("Visits", szVisits);
            hmProperty.put("Position", String.format(Locale.US, "%d", i));
            alAirports.add(hmProperty);

            childrenMaps.put(szKey, alAirports);
        }

        // put the above into arrayLists, but in the order that the keys were encountered.  .values() is an undefined order.
        ArrayList<HashMap<String, String>> headerList = new ArrayList<>();
        ArrayList<ArrayList<HashMap<String, String>>> childrenList = new ArrayList<>();
        for (String s : alKeys) {
            headerList.add(headers.get(s));
            childrenList.add(childrenMaps.get(s));
        }

        final SimpleExpandableListAdapter adapter = new SimpleExpandableListAdapter(
                getActivity(),
                headerList,
                R.layout.grouprow,
                new String[]{"sectionName"},
                new int[]{R.id.propertyGroup},
                childrenList,
                R.layout.visitedairportitem,
                new String[]{"Code", "Name", "Distance", "Visits"},
                new int[]{R.id.txtCode, R.id.txtName, R.id.txtDistance, R.id.txtVisits}
        );
        setListAdapter(adapter);

        // Auto-expand if 5 or fewer groups.
        boolean[] m_rgExpandedGroups = new boolean[alKeys.size()];
        if (m_rgExpandedGroups.length <= 5)
            for (int i = 0; i < m_rgExpandedGroups.length; i++)
                m_rgExpandedGroups[i] = true;

        for (int i = 0; i < m_rgExpandedGroups.length; i++)
            if (m_rgExpandedGroups[i])
                getExpandableListView().expandGroup(i);

        getExpandableListView().setOnChildClickListener((parent, v, groupPosition, childPosition, id) -> {

            if (!MFBMain.HasMaps())
                return false;

            @SuppressWarnings("unchecked")
            HashMap<String, String> hmProp = (HashMap<String, String>) adapter.getChild(groupPosition, childPosition);
            int position = Integer.parseInt(hmProp.get("Position"));
            String szRoute = "";
            String szAlias = "";
            if (position < 0) // all airports
                szRoute = VisitedAirport.toRoute(visitedAirports);
            else if (position < visitedAirports.length) {
                szRoute = visitedAirports[position].Code;
                szAlias = visitedAirports[position].Aliases;
            }

            Intent i = new Intent(ActVisitedAirports.this.getActivity(), ActFlightMap.class);
            i.putExtra(ActFlightMap.ROUTEFORFLIGHT, szRoute);
            i.putExtra(ActFlightMap.EXISTINGFLIGHTID, -1);
            i.putExtra(ActFlightMap.PENDINGFLIGHTID, -1);
            i.putExtra(ActFlightMap.ALIASES, szAlias);
            startActivity(i);

            return false;
        });
    }

    public void invalidate() {
        visitedAirports = null;
    }
}

File: app/src/main/java/com/myflightbook/android/ActVisitedAirports.kt
/*
	MyFlightbook for Android - provides native access to MyFlightbook
	pilot's logbook
    Copyright (C) 2017-2025 MyFlightbook, LLC

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package com.myflightbook.android

import android.content.Intent
import android.os.Bundle
import android.text.Editable
import android.text.TextWatcher
import android.view.*
import android.widget.EditText
import android.widget.ExpandableListView
import android.widget.SimpleExpandableListAdapter
import android.widget.TextView
import androidx.core.view.MenuHost
import androidx.core.view.MenuProvider
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.lifecycleScope
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout
import com.myflightbook.android.MFBMain.Invalidatable
import com.myflightbook.android.webservices.AuthToken
import com.myflightbook.android.webservices.MFBSoap.Companion.isOnline
import com.myflightbook.android.webservices.VisitedAirportSvc
import kotlinx.coroutines.launch
import model.MFBUtil.alert
import model.PackAndGo
import model.VisitedAirport
import model.VisitedAirport.Companion.toRoute
import java.text.DateFormat
import java.util.*

class ActVisitedAirports : ExpandableListFragment(), Invalidatable {
    private fun refreshVisitedAirports() {
        lifecycleScope.launch {
            ActMFBForm.doAsync<VisitedAirportSvc, Array<VisitedAirport>?>(requireActivity(),
                VisitedAirportSvc(),
                getString(R.string.prgVisitedAirports),
                { s -> s.getVisitedAirportsForUser(AuthToken.m_szAuthToken, requireContext()) },
                { _, result ->
                    if (result != null) {
                        visitedAirports = result
                        PackAndGo(requireContext()).updateAirports(visitedAirports)
                        populateList()
                    }
                }
            )
        }
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        return inflater.inflate(R.layout.expandablelist, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        // The usage of an interface lets you inject your own implementation
        val menuHost: MenuHost = requireActivity()

        // Add menu items without using the Fragment Menu APIs
        // Note how we can tie the MenuProvider to the viewLifecycleOwner
        // and an optional Lifecycle.State (here, RESUMED) to indicate when
        // the menu should be visible
        menuHost.addMenuProvider(object : MenuProvider {
            override fun onCreateMenu(menu: Menu, inflater: MenuInflater) {
                inflater.inflate(R.menu.currencymenu, menu)
            }

            override fun onMenuItemSelected(item: MenuItem): Boolean {
                if (item.itemId == R.id.menuRefresh) {
                    refreshAirports()
                    return true
                }
                return false
            }
        }, viewLifecycleOwner, Lifecycle.State.RESUMED)

        MFBMain.registerNotifyDataChange(this)
        MFBMain.registerNotifyResetAll(this)
        val v = requireView()
        val tvSearch = requireView().findViewById<TextView>(R.id.txtSearchProp)
        tvSearch.setHint(R.string.hintSearchVisitedAirports)
        tvSearch.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(charSequence: CharSequence, i: Int, i1: Int, i2: Int) {}
            override fun onTextChanged(charSequence: CharSequence, i: Int, i1: Int, i2: Int) {
                populateList()
            }

            override fun afterTextChanged(editable: Editable) {}
        })
        val srl: SwipeRefreshLayout = v.findViewById(R.id.swiperefresh)
        srl.setOnRefreshListener {
            refreshAirports()
        }
    }

    override fun onResume() {
        super.onResume()
        if (visitedAirports == null) refreshAirports() else populateList()
    }

    private fun refreshAirports() {
        if (isOnline(context)) refreshVisitedAirports() else {
            val p = PackAndGo(requireContext())
            val dt = p.lastAirportsPackDate()
            if (dt != null) {
                visitedAirports = p.cachedAirports()
                populateList()
                alert(
                    context,
                    getString(R.string.packAndGoOffline),
                    String.format(
                        Locale.getDefault(),
                        getString(R.string.packAndGoUsingCached),
                        DateFormat.getDateInstance().format(dt)
                    )
                )
            } else alert(context, getString(R.string.txtError), getString(R.string.errNoInternet))
        }
    }

    private fun populateList() {
        if (visitedAirports == null || visitedAirports!!.isEmpty()) return
        val rgap = visitedAirports!!

        // This maps the headers to the individual sub-lists.
        val headers = HashMap<String, HashMap<String, String>>()
        val childrenMaps = HashMap<String, ArrayList<HashMap<String, String>>>()

        // Keep a list of the keys in order
        val alKeys = ArrayList<String>()
        rgap.sort()
        visitedAirports = rgap // update to be sorted.

        // First do "All airports"
        var szKeyLast = getString(R.string.AllAirports)
        alKeys.add(szKeyLast)
        var hmAllAirports = HashMap<String, String>()
        hmAllAirports["sectionName"] = szKeyLast
        headers[szKeyLast] = hmAllAirports
        hmAllAirports = HashMap()
        hmAllAirports["Code"] = ""
        hmAllAirports["Name"] = szKeyLast
        val cVisits = visitedAirports!!.size
        hmAllAirports["Visits"] =
            resources.getQuantityString(R.plurals.uniqueAirportsCount, cVisits, cVisits)
        hmAllAirports["Distance"] = ""
        hmAllAirports["Position"] = "-1"
        val alAllAirports = ArrayList<HashMap<String, String>>()
        alAllAirports.add(hmAllAirports)
        childrenMaps[szKeyLast] = alAllAirports
        val szRestrict =
            (requireView().findViewById<View>(R.id.txtSearchProp) as EditText).text.toString().uppercase(Locale.getDefault())

        // slice and dice into headers/first names
        for (i in visitedAirports!!.indices) {
            val va = visitedAirports!![i]
            val szSearch = String.format(Locale.getDefault(), "%s %s %s %s", va.airport!!.facilityName, va.airport!!.airportID, va.airport!!.country ?: "", va.airport!!.admin1 ?: "").uppercase(Locale.getDefault()).trim()
            if (szRestrict.isNotEmpty() && !szSearch.contains(szRestrict))
                continue

            // get the first letter for this property as the grouping key
            val szKey = va.airport!!.facilityName[0].uppercase(Locale.getDefault())
            if (szKey.compareTo(szKeyLast) != 0) {
                alKeys.add(szKey)
                szKeyLast = szKey
            }
            val hmGroups: HashMap<String, String> =
                (if (headers.containsKey(szKey)) headers[szKey] else HashMap())!!
            hmGroups["sectionName"] = szKey
            headers[szKey] = hmGroups

            // Get the array-list for that key, creating it if necessary
            val alAirports: ArrayList<HashMap<String, String>> = if (childrenMaps.containsKey(szKey)) childrenMaps[szKey]!! else ArrayList()
            val hmProperty = HashMap<String, String>()
            val szDistance = if (va.airport!!.distance > 0) String.format(
                Locale.getDefault(),
                "(%.1fnm) ",
                va.airport!!.distance
            ) else ""
            val szEarliestVisit =
                DateFormat.getDateInstance(DateFormat.SHORT).format(va.earliestDate)
            val szLatestVisit = DateFormat.getDateInstance(DateFormat.SHORT).format(va.latestDate)
            val szVisits = if (va.numberOfVisits > 1) String.format(
                getString(R.string.vaMultiVisit),
                va.numberOfVisits,
                szEarliestVisit,
                szLatestVisit
            ) else String.format(getString(R.string.vaSingleVisit), szEarliestVisit)
            hmProperty["Code"] = String.format("%s - ", va.code)
            hmProperty["Name"] = va.airport!!.facilityName
            hmProperty["Distance"] = szDistance
            hmProperty["Visits"] = szVisits
            hmProperty["Position"] = String.format(Locale.US, "%d", i)
            alAirports.add(hmProperty)
            childrenMaps[szKey] = alAirports
        }

        // put the above into arrayLists, but in the order that the keys were encountered.  .values() is an undefined order.
        val headerList = ArrayList<HashMap<String, String>>()
        val childrenList = ArrayList<ArrayList<HashMap<String, String>>>()
        for (s in alKeys) {
            headerList.add(headers[s]!!)
            childrenList.add(childrenMaps[s]!!)
        }
        val adapter = SimpleExpandableListAdapter(
            activity,
            headerList,
            R.layout.grouprow,
            arrayOf("sectionName"),
            intArrayOf(R.id.propertyGroup),
            childrenList,
            R.layout.visitedairportitem,
            arrayOf("Code", "Name", "Distance", "Visits"),
            intArrayOf(R.id.txtCode, R.id.txtName, R.id.txtDistance, R.id.txtVisits)
        )
        listAdapter = adapter

        // Auto-expand if 5 or fewer groups.
        val mRgexpandedgroups = BooleanArray(alKeys.size)
        if (mRgexpandedgroups.size <= 5) Arrays.fill(mRgexpandedgroups, true)
        for (i in mRgexpandedgroups.indices) if (mRgexpandedgroups[i]) expandableListView!!.expandGroup(
            i
        )
        expandableListView!!.setOnChildClickListener { _: ExpandableListView?, _: View?, groupPosition: Int, childPosition: Int, _: Long ->
            if (!MFBMain.hasMaps() || !isOnline(
                    context
                )
            ) return@setOnChildClickListener false
            val hmProp = adapter.getChild(groupPosition, childPosition) as HashMap<*, *>
            val position = (hmProp["Position"]!! as String).toInt()
            var szRoute = ""
            var szAlias: String? = ""
            if (position < 0 && visitedAirports != null) // all airports
                szRoute = toRoute(visitedAirports!!) else if (position < visitedAirports!!.size) {
                szRoute = visitedAirports!![position].code
                szAlias = visitedAirports!![position].aliases
            }
            val i = Intent(this@ActVisitedAirports.activity, ActFlightMap::class.java)
            i.putExtra(ActFlightMap.ROUTEFORFLIGHT, szRoute)
            i.putExtra(ActFlightMap.EXISTINGFLIGHTID, -1)
            i.putExtra(ActFlightMap.PENDINGFLIGHTID, -1L)
            i.putExtra(ActFlightMap.ALIASES, szAlias)
            startActivity(i)
            false
        }
    }

    override fun invalidate() {
        visitedAirports = null
    }

    companion object {
        private var visitedAirports: Array<VisitedAirport>? = null
    }
}